File: PythonCode/Chapter3_Time-indepdentSE/Exercise_3_1_3.py
""" 
 This script determines the admissible energies for a hydrogen atom.
 It does so by setting up the Hamiltonian, numerically, and 
 diagonalizing it. It does so by using a three point finite difference
 formula. The boundary conditions at r = 0 and r = L, where L is the size
 of the numerical domain, is included manifestly in our approximation.

 In addition to determining the bound energies, the script also plots the
 admissible energies and compares the spectrum with the Bohr formula.

 Numerical inputs:
   L       - The extension of the spatial grid 
   N       - The number of grid points
 
 The number of wave functions to plot: Nplot
 
 All inputs are hard coded initially.
"""

# Libraries
import numpy as np
from matplotlib import pyplot as plt


# Numerical grid parameters
L = 50;
N = 1000;                

# The number of plots
Nplots = 8;

# Set up grid. Here r = 0 and r = L are excluded. It does, however, use
# that psi(0) = psi(L) = 0.
h = L/(N+1);
r = np.linspace(h, L-h, N);      
# Convert to column vector
r = np.transpose(r)

# Set up potential
Vpot = -1/r

# Set up matrix for the kinetic energy - 3 point formula
# Allocate and declare
Tmat_FD3 = np.zeros((N, N), dtype=complex)
# Endpoints
Tmat_FD3[0, :2] = [-2, 1]
Tmat_FD3[N-1, (N-2):] = [1, -2]
# Interior points
for n in range (1,N-1):
  Tmat_FD3[n, [n-1, n, n+1]] = [1, -2, 1]
# Correct pre-factors
Tmat_FD3 = Tmat_FD3/h**2
Tmat_FD3 = -1/2*Tmat_FD3  



# Full Hamiltonian
Ham = Tmat_FD3 + np.diag(Vpot);         

# Diagaonalize Hamiltonian (Hermitian matrix)
Evector, PsiMat = np.linalg.eigh(Ham)
# Normalize eigenstates
PsiMat = PsiMat/np.sqrt(h)

# Plot eigenfunctions and write bound energies to screen
plt.figure(1)
plt.clf()
n = 0
while Evector[n] < 0:
    print(f'Admissible bound energy: {Evector[n]:.4f}')
    LegendEntry = r'$\Psi_n$ for n = ' + str(n+1)
    plt.plot(r, PsiMat[:,n], label = LegendEntry)
    n = n+1

# Insert legend and grid
plt.grid()
plt.xlabel('r [a. u.]', fontsize = 12)
plt.legend()
plt.show()

# Plot comparison to the Bohr formula
Nbound = n
plt.figure(2)
plt.clf()
plt.plot(range(1,Nbound+1), np.real(Evector[:Nbound]), 'x', 
         color = 'red', label = 'Numerical')    
plt.plot(range(1,Nbound+1), -0.5/np.linspace(1, Nbound, Nbound)**2, 
         'o', color = 'black', mfc = 'none', label = 'Bohr formula')
plt.grid()
plt.xlabel('Quantum number n', fontsize = 12)
plt.ylabel('Energy [a.u.]', fontsize = 12)
plt.legend()
plt.show()

File: MatlabCode/Chapter3_Time-independentISE/Exercise_3_1_3.m
% This script determines the admissible energies for a hydrogen atom.
% It does so by setting up the Hamiltonian, numerically, and 
% diagonalizing it. The kinetic energy operator is estimated using
% a three point finite difference formula. The boundary conditions at 
% r = 0 and r = L, where L is the size of the numerical domain, is 
% included manifestly in our approximation.
%
% In addition to determining the bound energies, the script also plots the
% admissible energies and compares the spectrum with the Bohr formula.
%
% Numerical inputs:
% L       - The extension of the spatial grid 
% N       - The number of grid points
% 
% Nplot   - The number of wave functions to plot
% 
% All inputs are hard coded initially.

% Numerical grid parameters
L = 50;
N = 1000;                

% The number of plots
Nplots = 8;

% Set up grid. Here r = 0 and r = L are excluded. It does, however, use
% that psi(0) = psi(L) = 0.
h = L/(N+1);
r = transpose(linspace(h, L-h, N));      % Column vector

% Set up potetial
Vpot = -1./r;

% Set up kinetic energy matrix by means of three point-formula
Tmat_FD3 = zeros(N, N);
Tmat_FD3(1, [1 2]) = [-2 1];
Tmat_FD3(N, [N-1 N]) = [1 -2];
for n=2:(N-1)
  Tmat_FD3(n, [n-1 n n+1]) = [1 -2 1]; 
end
Tmat_FD3 = Tmat_FD3/h^2;        % Correct duble derivative
Tmat_FD3 = -1/2*Tmat_FD3;       % Correct prefactor

% Full Hamiltonian
Ham = Tmat_FD3 + diag(Vpot);         % Hamiltonian

% Diagonalization
[U E] = eig(Ham);
% Extract eigen energies, the diagonal of E
E = diag(E);
% Sort eigenvalues - and rearrange eigenvectors accordingly
[E Ind] = sort(E);
U = U(:, Ind);
% Ensure proper normalization
U = U/sqrt(h);

% Determine the number of bound states
Nbound = length(find(E<0));
disp(['We found ', num2str(Nbound), ' numerical bound states.'])
% Plott the bound energies together with the Bohr formula
figure(1)
for n = 1:Nbound
  % Numerical values
  plot(1:Nbound, E(1:Nbound), 'rx', 'linewidth', 2)
  % Analytical values
  hold on
  plot(1:Nbound, -1./(2*[1:Nbound].^2), 'ko', 'linewidth', 2)
  hold off
end
grid on
set(gca, 'fontsize', 15)
xlabel('n quantum number')
ylabel('Energy')
legend('Numerical eigen energies', 'Bohr formula', 'location', 'southeast')

% Plot some of the wave functions
figure(2)
for n = 1:min(N,Nplots)
  % Write energy to screen
  plot(r, U(:,n), 'k-', 'linewidth', 2)
  title(['Energy index: ', num2str(n), ', energy: ', num2str(E(n))])
  set(gca, 'fontsize', 12)
  xlabel('r [a.u.]')
  ylabel('Wave function')
  grid on
  pause
end